IPC::StreamClientConnection should support sendWithAsyncReply()

kkinnunen-apple opened the pull request:
#### 19a30ab35aa9d8d0ebbe7b1cd002b5573c306e81
<pre>
IPC::StreamClientConnection should support sendWithAsyncReply()
<a href="https://bugs.webkit.org/show_bug.cgi?id=235965">https://bugs.webkit.org/show_bug.cgi?id=235965</a>
rdar://problem/88636456&gt;

Reviewed by Simon Fraser and Matt Woodrow.

Implement the sending async messages by:
- Sending the message as stream message if possible
  - Otherwise send as &quot;out of stream&quot; message
- Reply comes through the dedicated Connection

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::addAsyncReplyHandler):
(IPC::Connection::sendMessageWithAsyncReply):
(IPC::Connection::dispatchMessage):
(IPC::CompletionHandler&lt;void):
(IPC::addAsyncReplyHandler): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::waitForAsyncCallbackAndDispatchImmediately):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::send):
(IPC::StreamClientConnection::trySendStream):
(IPC::StreamClientConnection::sendWithAsyncReply): Deleted.
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::StreamServerConnection::sendAsyncReply):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;Namespace::Subnamespace::StructName&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::OtherClass&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::ReturnRefClass&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::InheritsFrom&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::InheritanceGrandchild&gt;::decode):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::jsValueForReplyArguments):
(IPC::messageArgumentDescriptions):
(IPC::messageReplyArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
(IPC::description):
(IPC::receiverName):
(IPC::messageAllowedWhenWaitingForSyncReply):
(WTF::void&gt;):
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStream.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessageReceiver.cpp:
(WebKit::TestWithStream::didReceiveStreamMessage):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithStream_SendStringAsync&gt;):
(IPC::jsValueForDecodedMessageReply&lt;MessageName::TestWithStream_SendStringAsync&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithStream_SendStringSync&gt;):
(IPC::jsValueForDecodedMessageReply&lt;MessageName::TestWithStream_SendStringSync&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithStream_SendStringSynchronized&gt;): Deleted.
(IPC::jsValueForDecodedMessageReply&lt;MessageName::TestWithStream_SendStringSynchronized&gt;): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
(Messages::TestWithStream::SendStringAsync::name):
(Messages::TestWithStream::SendStringAsync::asyncMessageReplyName):
(Messages::TestWithStream::SendStringAsync::SendStringAsync):
(Messages::TestWithStream::SendStringAsync::arguments const):
(Messages::TestWithStream::SendStringSync::name):
(Messages::TestWithStream::SendStringSync::SendStringSync):
(Messages::TestWithStream::SendStringSynchronized::name): Deleted.
(Messages::TestWithStream::SendStringSynchronized::asyncMessageReplyName): Deleted.
(Messages::TestWithStream::SendStringSynchronized::SendStringSynchronized): Deleted.
(Messages::TestWithStream::SendStringSynchronized::arguments const): Deleted.
(Messages::TestWithStream::ReceiveMachSendRight::asyncMessageReplyName): Deleted.
(Messages::TestWithStream::SendAndReceiveMachSendRight::asyncMessageReplyName): Deleted.
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(IPC::StreamClientConnection::sendWithAsyncReply):
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/257735@main">https://commits.webkit.org/257735@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de41fbf9cc1026d23655ac02bc8ac54e4f608f81

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99840 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9016 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32924 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109200 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169438 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103842 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9884 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86309 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92300 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107110 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105607 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90747 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34201 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/103351 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22136 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2829 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23646 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2777 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/46025 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5314 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8912 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43119 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4637 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -179,11 +184,44 @@ bool StreamClientConnection::send(T&& message, ObjectIdentifier<U> destinationID
+        return replyID;
+    // replyHandlerToCancel might be already cancelled if invalidate() happened in-between.
+    if (auto replyHandlerToCancel = m_connection->takeAsyncReplyHandler(replyID)) {
+        // FIXME: Current contract is that completionHandler is called on the connection run loop.
```
Does this need a bug, referenced by the comment?

smfr commented on Source/WebKit/Platform/IPC/StreamServerConnection.h:
```diff
@@ -82,6 +82,9 @@ class StreamServerConnection final : public ThreadSafeRefCounted<StreamServerCon
     void sendSyncReply(Connection::SyncRequestID, Arguments&&...);
 
+    template<typename T, typename... Arguments>
+    void sendAsyncReply(uint64_t asyncReplyID, Arguments&&...);
```
Please make a type for the `asyncReplyID` like `Connection::SyncRequestID`

kkinnunen-apple committed:
IPC::StreamClientConnection should support sendWithAsyncReply()
https://bugs.webkit.org/show_bug.cgi?id=235965
rdar://problem/88636456>

Reviewed by Simon Fraser and Matt Woodrow.

Implement the sending async messages by:
- Sending the message as stream message if possible
  - Otherwise send as "out of stream" message
- Reply comes through the dedicated Connection

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::addAsyncReplyHandler):
(IPC::Connection::sendMessageWithAsyncReply):
(IPC::Connection::dispatchMessage):
(IPC::CompletionHandler<void):
(IPC::addAsyncReplyHandler): Deleted.
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::waitForAsyncCallbackAndDispatchImmediately):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::send):
(IPC::StreamClientConnection::trySendStream):
(IPC::StreamClientConnection::sendWithAsyncReply): Deleted.
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::StreamServerConnection::sendAsyncReply):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<Namespace::Subnamespace::StructName>::decode):
(IPC::ArgumentCoder<Namespace::OtherClass>::decode):
(IPC::ArgumentCoder<Namespace::ReturnRefClass>::decode):
(IPC::ArgumentCoder<WebCore::InheritsFrom>::decode):
(IPC::ArgumentCoder<WebCore::InheritanceGrandchild>::decode):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::jsValueForReplyArguments):
(IPC::messageArgumentDescriptions):
(IPC::messageReplyArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
(IPC::description):
(IPC::receiverName):
(IPC::messageAllowedWhenWaitingForSyncReply):
(WTF::void>):
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStream.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessageReceiver.cpp:
(WebKit::TestWithStream::didReceiveStreamMessage):
(IPC::jsValueForDecodedMessage<MessageName::TestWithStream_SendStringAsync>):
(IPC::jsValueForDecodedMessageReply<MessageName::TestWithStream_SendStringAsync>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithStream_SendStringSync>):
(IPC::jsValueForDecodedMessageReply<MessageName::TestWithStream_SendStringSync>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithStream_SendStringSynchronized>): Deleted.
(IPC::jsValueForDecodedMessageReply<MessageName::TestWithStream_SendStringSynchronized>): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
(Messages::TestWithStream::SendStringAsync::name):
(Messages::TestWithStream::SendStringAsync::asyncMessageReplyName):
(Messages::TestWithStream::SendStringAsync::SendStringAsync):
(Messages::TestWithStream::SendStringAsync::arguments const):
(Messages::TestWithStream::SendStringSync::name):
(Messages::TestWithStream::SendStringSync::SendStringSync):
(Messages::TestWithStream::SendStringSynchronized::name): Deleted.
(Messages::TestWithStream::SendStringSynchronized::asyncMessageReplyName): Deleted.
(Messages::TestWithStream::SendStringSynchronized::SendStringSynchronized): Deleted.
(Messages::TestWithStream::SendStringSynchronized::arguments const): Deleted.
(Messages::TestWithStream::ReceiveMachSendRight::asyncMessageReplyName): Deleted.
(Messages::TestWithStream::SendAndReceiveMachSendRight::asyncMessageReplyName): Deleted.
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(IPC::StreamClientConnection::sendWithAsyncReply):
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/257735@main
